### ISSUE TITLE
Support for custom CA certificates in OpenID authentication

### ISSUE THREAD

**Issue by AbrohamLincoln** [kiali/kiali#4050]
**Is your feature request related to a problem? Please describe.**
OpenID authentication does not support custom CA certificates.  The only recourse is to disable TLS verification.

**Describe the solution you'd like**
Add the ability to add in a custom root CA to the OpenID config.

Do something similar to [this](https://github.com/kiali/kiali/blob/master/business/openshift_oauth.go#L248-L257) in the [OpenID http client](https://github.com/kiali/kiali/blob/5bcdd2f372ff2fcca4746e321bd9e93e677e1b0c/business/openid_auth.go#L778)

**Describe alternatives you've considered**
Supporting custom CAs could potentially solve this as well: https://github.com/kiali/kiali/issues/1675

**Comment by jshaughn** [comment]
cc @israel-hdez

**Comment by israel-hdez** [comment]
@AbrohamLincoln I wonder if you can workaround this requirement by building a customized Kiali image with the needed certificates?

For #1675, Kiali would need a writable file-system in order to update the trusted certificate chains. We [recently added security constraints that run Kiali with a read-only file system](https://github.com/kiali/kiali-operator/blob/f0c07d1323d67b392d29e46073d81e14d032495e/roles/default/kiali-deploy/templates/kubernetes/deployment.yaml#L54-L58). So, #1675 is blocked for now.

However, if you are able to build and use a customized Kiali image, additional certificates can be added. A sample Dockerfile is (I haven't tried, but this should be a good start point):

```Dockerfile
FROM quay.io/kiali/kiali:latest

USER root

COPY your_ca_file /etc/pki/ca-trust/source/anchors/
RUN /bin/update-ca-trust

USER 1000
```

I would like to avoid a new ca_file option, because we have such options for almost every single service that Kiali talks to (see, for example, [this option for prometheus](https://github.com/kiali/kiali-operator/blob/f0c07d1323d67b392d29e46073d81e14d032495e/deploy/kiali/kiali_cr.yaml#L588-L590)). I think this way is simply adding too much config properties and also support in the code is required. So, I want to avoid a new one. Adding the needed certs to the image is, IMO, a cleaner solution.

If you can't use a custom image, I think we can do the effort of adding yet another one `ca_file` setting. Similarly to the OpenShift auth, only a single cert file would be supported.

**Comment by AbrohamLincoln** [comment]
Custom images are technically possible, but more or less a non-starter policy wise in our environment.

What about adding an initContainer, a volumeMount, and an emptyDir volume to the Kiali deployment managed by the operator?  The initContainer creates the trust store on the emptyDir volume which includes any custom CA certificates provided as a configMap.  Then Kiali mounts the same volume in?

**Comment by jmazzitelli** [comment]
> What about adding an initContainer, a volumeMount, and an emptyDir volume to the Kiali deployment managed by the operator? The initContainer creates the trust store on the emptyDir volume which includes any custom CA certificates provided as a configMap. Then Kiali mounts the same volume in?

I would prefer to just have a new single ca_cert setting in the auth config rather than go through that mess :)

**Comment by israel-hdez** [comment]
> What about adding an initContainer, a volumeMount, and an emptyDir volume to the Kiali deployment managed by the operator? The initContainer creates the trust store on the emptyDir volume which includes any custom CA certificates provided as a configMap. Then Kiali mounts the same volume in?

I agree with @jmazzitelli. This solution would be more complex. So, I'll go ahead with adding a new ca_file attribute in the Kiali CR to add the desired custom certificate. As I said, only a single certificate is going to be supported.

**Comment by israel-hdez** [comment]
This is done. I'm closing.
This will be available in Kiali v1.37.

In case you want to try a snapshot or an edge build, here are the staging docs: https://kiali.io/documentation/staging/configuration/authentication/openid/#_using_an_openid_provider_with_a_self_signed_certificate

**Comment by AbrohamLincoln** [comment]
Thanks!  Much appreciated.
